components/places/sql/create_shared_triggers.sql

```diff
@@ -261,6 +268,20 @@ BEGIN
     WHERE id = OLD.placeId;
 END;
 
+-- Similar to cleanup_pages, if the origin/place remains with no foreign references
+-- and no visits it should be deleted.
+-- This approach may not be suitable for desktop but seems to be for us - see
+-- https://bugzilla.mozilla.org/show_bug.cgi?id=1650511#c41 for more discussion.
+CREATE TEMP TRIGGER moz_cleanup_origin_bookmark_deleted_trigger
+AFTER DELETE ON moz_bookmarks
+BEGIN
+    DELETE FROM moz_places
+        WHERE id = OLD.fk
+        AND foreign_count = 0
+        AND last_visit_date_local = 0
+        AND last_visit_date_remote = 0;
```

Comment:
I wonder if we should do the equivalent of `delete_pending_temp_tables` here—or at least a `DELETE FROM moz_updateoriginsdelete_temp`.

I think that would mean the origin would get cleaned up as soon as the bookmark is deleted, and you could avoid the call to `crate::storage::history::delete_everything(&conn)?` in the bookmarks test.

WDYT?

Reply:
Oh, rad!

> the trigger on `moz_bookmarks` causes a deletion of `moz_places`, which causes the `moz_places_afterdelete_trigger_origins` trigger.

Yep, you're right!

But the `moz_places_afterdelete_trigger_origins` trigger just inserts a row for the soon-to-be-deleted origin into `moz_updateoriginsdelete_temp`. It doesn't actually delete the origin from `moz_origins`, or update the global frecency stats in `moz_meta`, until you do `DELETE FROM moz_updateoriginsdelete_temp` to fire the `moz_updateoriginsdelete_afterdelete_trigger`.

At first, I thought this would leave the origin behind, because doing `DELETE FROM moz_places` by itself isn't enough to force the `delete_pending_temp_tables` dance. And that's why I thought we needed to force it in the trigger.

But then I realized that `delete_bookmark` calls `delete_pending_temp_tables`! So it's not necessary for your new trigger to do that, because the `delete_pending_temp_tables` call [here](https://github.com/mozilla/application-services/blob/e0b1d5b47b781bb420f6e6fff3bd0ddae408e5cc/components/places/src/storage/bookmarks.rs#L452) will do that work already.

So this is all working correctly now—if in a slightly indirect way that I didn't think all the way through—and you can ignore my suggestion! 😅